src/sci/impl/reflector.cljc

```diff
@@ -0,0 +1,326 @@
+       (or (identical? arg-type Double)
+           (identical? arg-type Float/TYPE))
+
+       (identical? param-type Long/TYPE)
+       (or (identical? arg-type Long)
+           (identical? arg-type Integer/TYPE)
+           (identical? arg-type Short/TYPE)
+           (identical? arg-type Byte/TYPE))
+
+       (identical? param-type Character/TYPE)
+       (identical? arg-type Character)
+
+       (identical? param-type Short/TYPE)
+       (identical? arg-type Short)
+
+       (identical? param-type Byte/TYPE)
+       (identical? arg-type Byte)
+
+       (identical? param-type Boolean/TYPE)
+       (identical? arg-type Boolean)
+
+       :else
+       false)))
+
+#?(:clj
+   (defn- is-congruent?
+     "Check if parameters are congruent with arguments and arg-types."
+     [^objects params ^objects args ^objects arg-types]
+     (if (nil? args)
+       (zero? (alength params))
+       (and (= (alength params) (alength args))
+            (loop [i 0]
+              (if (< i (alength params))
+                (let [arg (aget args i)
+                      arg-type (if arg-types
+                                 (let [t (aget arg-types i)]
+                                   (if (and (nil? t) (some? arg))
+                                     (.getClass ^Object arg)
+                                     t))
+                                 (when arg (.getClass ^Object arg)))
+                      param-type (aget params i)]
+                  (if (param-arg-type-match? param-type arg-type)
+                    (recur (inc i))
+                    false))
+                true))))))
+
+#?(:clj
+   (defn- match-method
+     "Find the best matching method from a list of methods given args and optional arg-types."
+     [methods args arg-types]
+     (loop [methods methods
+            found-m nil]
+       (if-let [^Method m (first methods)]
+         (let [params (.getParameterTypes m)]
+           (if (and (is-congruent? params args arg-types)
+                    (or (nil? found-m)
+                        (Compiler/subsumes params (.getParameterTypes ^Method found-m))))
+             (recur (rest methods) m)
+             (recur (rest methods) found-m)))
+         found-m))))
+
+#?(:clj
+   (defn- widen-boxed-args
+     "Widen boxed numeric arguments (e.g., Integer -> Long, Float -> Double)."
+     [^objects args]
+     (let [widened (object-array (alength args))]
+       (dotimes [i (alength args)]
+         (let [arg (aget args i)]
+           (if (some? arg)
+             (let [val-class (.getClass ^Object arg)]
+               (aset widened i
+                     (cond
+                       (or (identical? val-class Integer) (identical? val-class Short) (identical? val-class Byte))
+                       (.longValue ^Number arg)
+
+                       (identical? val-class Float)
+                       (.doubleValue ^Number arg)
+
+                       :else
+                       arg)))
+             (aset widened i nil))))
```

Comment:
`(aset widened i nil)` wasn't in the original code. Possibly redundant depending on the semantics of `(object-array NUMBER)`.

Reply:
applied